File: 2_Creational_Design_Patterns/2.01_Builder/Python_Snippet/Builder.py
'''
Create a Pizza maker and allow options to choose Base, Sauce and Topping
'''

class PizzaMaker():
    def __init__(self,
                 is_thin_crust=False, is_pan_crust=False,
                 has_tomato_sauce=False,has_bbq_sauce=False,
                 has_tomato=False,has_mushrooms=False,has_pepperoni=False) -> None:
        self.base = None
        self.sauce = None
        self.toppings = []
        
        # set pizza base
        if is_thin_crust:
            self.base = 'thin crust'
        else:
            self.base = 'pan crust'
        
        # set pizza sauce
        if has_tomato_sauce:
            self.sauce = 'tomato'
        else:
            self.sauce = 'bbq'
        
        # set toppings
        if has_tomato:
            self.toppings.append('tomato')
        if has_mushrooms:
            self.toppings.append('mushrooms')
        if has_pepperoni:
            self.toppings.append('pepperoni')
    def __str__(self) -> str:
        return f"Pizza with {self.base} base, {self.sauce} sauce and {', '.join(self.toppings)} toppings"

# with to add more toppings the number of parameters will increase 
# we use builder to solve this 

class Pizza():
    def __init__(self,base,sauce,toppings) -> None:
        self.base = base
        self.sauce = sauce
        self.toppings = toppings
    
    def __str__(self) -> str:
        return f"Pizza with {self.base} base, {self.sauce} sauce and {', '.join(self.toppings)} toppings"
    
class PizzaBuilder():
    def __init__(self) -> None:
        self.base = None
        self.sauce = None
        self.toppings = []
    
    def set_base(self,base):
        self.base = base
        return self
    
    def set_sauce(self,sauce):
        self.sauce = sauce
        return self
    
    def add_topping(self,topping): 
        self.toppings.append(topping)
        return self
    
    def build(self):
        if not all([self.base,self.sauce]):
            raise Exception('Pizza must have base and sauce')
        return Pizza(self.base,self.sauce,self.toppings)

pizzaBuilder = PizzaBuilder()
veg_pizza = pizzaBuilder.set_base('Thin Crust').set_sauce('Tomato').add_topping('Olive').add_topping('Tomato').add_topping('Cheese').add_topping('pepper').build()
print(veg_pizza)
    

File: 2_Creational_Design_Patterns/2.01_Builder/Java_Snippet/Builder.java

// Create a Pizza maker and allow options to choose Base, Sauce and Topping
import java.util.*;

class PizzaMaker {
        private String base;
        private String sauce;
        private List<String> toppings;

        public PizzaMaker(boolean isThinCrust, boolean isPanCrust,
                        boolean hasTomatoSauce, boolean hasBbqSauce,
                        boolean hasTomato, boolean hasPepper, boolean hasMushrooms) {
                this.toppings = new ArrayList<String>();
                if (isThinCrust)
                        this.base = "thin crust";
                if (isPanCrust)
                        this.base = "pan crust";
                if (hasTomatoSauce)
                        this.sauce = "tomato";
                if (hasBbqSauce)
                        this.sauce = "bbq";
                if (hasTomato)
                        this.toppings.add("tomato");
                if (hasPepper)
                        this.toppings.add("pepper");
                if (hasMushrooms)
                        this.toppings.add("mushroom");
        }

        @java.lang.Override
        public java.lang.String toString() {
                return String.format("Pizza with %s base, %s sauce and %s toppings", base, sauce,
                                String.join(", ", toppings));
        }

        public static void main(String[] args) {
                PizzaMaker pizzaMaker = new PizzaMaker(true, false, true, false, true, true, false);
                System.out.println(pizzaMaker);
        }
}

// with to add more toppings the number of parameters will increase
// we use builder to solve this
class Pizza {

        private final String base;
        private final String sauce;
        private final List<String> toppings;

        private Pizza(String base, String sauce, List<String> toppings) {
                this.base = base;
                this.sauce = sauce;
                this.toppings = toppings;
        }

        @Override
        public String toString() {
                return String.format("Pizza with %s base, %s sauce and %s toppings", base, sauce,
                                String.join(", ", toppings));
        }

        public static class PizzaBuilder {

                private String base;
                private String sauce;
                private final List<String> toppings = new ArrayList<>();

                public PizzaBuilder setBase(String base) {
                        this.base = base;
                        return this;
                }

                public PizzaBuilder setSauce(String sauce) {
                        this.sauce = sauce;
                        return this;
                }

                public PizzaBuilder addTopping(String topping) {
                        toppings.add(topping);
                        return this;
                }

                public Pizza build() {
                        if (base == null || sauce == null) {
                                throw new IllegalArgumentException("Pizza must have base and sauce!");
                        }
                        return new Pizza(base, sauce, toppings);
                }
        }

        public static void main(String[] args) {
                PizzaBuilder pizzaBuilder = new PizzaBuilder();
                Pizza vegPizza = pizzaBuilder.setBase("Thin Crust")
                                .setSauce("Tomato")
                                .addTopping("Olive")
                                .addTopping("Tomato")
                                .addTopping("Cheese")
                                .addTopping("Pepper")
                                .build();

                System.out.println(vegPizza);
        }
}